### PR TITLE
Add a queue for checkpoint read

### PR DESCRIPTION
**Pull request by kaituo**
Note: since there are a lot of dependencies, I only list the main class and test code to save reviewers' time. The build will fail due to missing dependencies. I will use that PR just for review. will not merge it. Will have a big one in the end and merge once after all review PRs get approved. Now the code is missing unit tests. Posting PRs now to meet the cutoff date (June 1). Will add unit tests, run performance tests, and fix bugs before the official release.

### Description
This PR adds a queue for loading model checkpoint.  The read is a multi-get query. Possible results are:
* If a checkpoint is not found, we forward that request to cold start queue.
*When a request gets errors, the queue does not change its expiry time and puts that request to the end of the queue and automatically retries them before they expire.
* When a checkpoint is found, we load that point to memory and score the input data point and save the result if a complete model exists. Otherwise, we enqueue the sample.  If we can host that model in memory (e.g., there is enough memory), we put the loaded model to cache.  Otherwise (e.g., a cold entity), we write the updated checkpoint back to disk.

Testing done:
1. Manual tests using 10 HCAD detectors and 12,000 entities in a 3 node cluster.

 
### Check List
- [ Y ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+                // retry all of them
+                super.putAll(toProcess);
+            } else {
+                LOG.error("Fail to restore models", exception);
+            }
+        });
+    }
+
+    private void processCheckpointIteration(
+        int i,
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java:227]: I think this name be more descriptive.

File: src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+            if (checkpointResponse != null) {
+                // successful requests
+                Optional<Map<String, Object>> checkpointString = checkpointDao.processRawCheckpoint(checkpointResponse.getResponse());
+                if (checkpointString.isPresent()) {
+                    // checkpoint exists
+                    Optional<Entry<EntityModel, Instant>> checkpoint = checkpointDao
+                        .fromEntityModelCheckpoint(checkpointString.get(), modelId);
+
+                    if (false == checkpoint.isPresent()) {
+                        // checkpoint is too big
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java:263]: Is this the only case where the checkpoint isnt present?

File: src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+
+    @Override
+    protected MultiGetRequest toBatchRequest(List<EntityFeatureRequest> toProcess) {
+        MultiGetRequest multiGetRequest = new MultiGetRequest();
+        for (EntityRequest request : toProcess) {
+            Optional<String> modelId = request.getModelId();
+            if (false == modelId.isPresent()) {
+                continue;
+            }
+            multiGetRequest.add(new MultiGetRequest.Item(CommonName.CHECKPOINT_INDEX_NAME, modelId.get()));
```
[the remark this answers is not in the document]
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java:159]: It is imposed here: https://github.com/kaituo/anomaly-detection-1/blob/multiCat4/src/main/java/org/opensearch/ad/ratelimit/RateLimitedRequestWorker.java#L387

So toBatchRequest method can take the input and send the multi-get directly.

File: src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+                        // submit to cold start queue
+                        entityColdStartQueue.put(origRequest);
+                    }
+                }
+            }
+
+            processCheckpointIteration(0, toProcess, successfulRequests, retryableRequests);
+        }, exception -> {
+            if (ExceptionUtil.isOverloaded(exception)) {
+                LOG.error("too many get AD model checkpoint requests or shard not avialble");
```
[the remark this answers is not in the document]
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java:215]: I am checking the exception is UnavailableShardsException.  I observed that can happen when the system cannot respond to requests in the performance tests.

[the remark this answers is not in the document]
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java:215]: yes, fixed avialble.

File: src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+                setCoolDownStart();
+            } else if (ExceptionUtil.isRetryAble(exception)) {
+                // retry all of them
+                super.putAll(toProcess);
```
[the remark this answers is not in the document]
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java:219]: yeah, removed

File: src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java (hunk 1 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java:227]: Please check https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

"Note that certain universally-known abbreviations are OK, such as i for an iteration variable and T for a template parameter."

File: src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java (hunk 2 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java:263]: yes it is so far

**Review by jmazanec15 — APPROVED** [review]

File: src/main/java/org/opensearch/ad/ratelimit/CheckpointReadWorker.java [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+            for (MultiGetItemResponse itemResponse : itemResponses) {
+                String modelId = itemResponse.getId();
+                if (itemResponse.isFailed()) {
+                    final Exception failure = itemResponse.getFailure().getFailure();
+                    if (failure instanceof IndexNotFoundException) {
+                        for (EntityRequest origRequest : toProcess) {
+                            // submit to cold start queue
+                            entityColdStartQueue.put(origRequest);
+                        }
+                        return;
```
Comment (jngz-es) [inline src/main/java/org/opensearch/ad/ratelimit/CheckpointReadWorker.java:189]: I am not quite sure, is it return? not continue? I saw there is for loop outside.

Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/CheckpointReadWorker.java:189]: It is return.  If it is checkpoint index not found exception, I don't need to retry as checkpoint read is bound to fail. Just send everything to the cold start queue and return. Added comments to explain.

**Review by jngz-es — APPROVED** [review]

**Pull request closed without merging**